File: game-server/src/main/java/org/forfun/mmorpg/game/account/AccountService.java
package org.forfun.mmorpg.game.account;

import org.forfun.mmorpg.game.base.EntityCacheService;
import org.forfun.mmorpg.game.base.GameContext;
import org.forfun.mmorpg.game.database.user.dao.AccountDao;
import org.forfun.mmorpg.game.database.user.entity.AccountEnt;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.cache.annotation.CachePut;
import org.springframework.cache.annotation.Cacheable;
import org.springframework.stereotype.Service;

@Service
public class AccountService implements EntityCacheService<AccountEnt, Long> {

	@Autowired
	private AccountDao accountDao;

	public AccountEnt createNew(AccountEnt account) {
		return putEntity(account);
	}

	@Override
	@Cacheable(cacheNames = "account")
	public AccountEnt getEntity(Long id) {
		return accountDao.findById(id).orElse(null);
	}

	@Override
	@CachePut(cacheNames = "account")
	public AccountEnt putEntity(AccountEnt account) {
		GameContext.getPlayerService().addAccountProfile(account);
		GameContext.getAysncDbService().saveToDb(account);
		return account;
	}
}

File: game-server/src/main/kotlin/org/forfun/mmorpg/game/account/AccountService.kt
package org.forfun.mmorpg.game.account

import org.forfun.mmorpg.game.database.user.entity.AccountEnt;
import org.forfun.mmorpg.framework.cache.BaseEntityCacheService
import org.forfun.mmorpg.framework.cache.EntityCacheAutowired
import org.springframework.stereotype.Service;
import kotlin.jvm.optionals.getOrNull

@Service
class AccountService {


    @EntityCacheAutowired
    private lateinit var accountCacheService: BaseEntityCacheService<AccountEnt, Long>

    fun createNew(account: AccountEnt): AccountEnt {
        return accountCacheService.getOrCreate(account.id) {
            var account = AccountEnt()
            account
        };
    }

    fun getEntity(id:Long):AccountEnt? {
        return accountCacheService.getEntity(id).getOrNull()
    }

}
